apps/dash-tsne/demo.py

```diff
@@ -0,0 +1,624 @@
+                                    ),
+                                ]
+                            )
+                        ],
+                    ),
+                    html.Div(
+                        className="six columns",
+                        children=[
+                            dcc.Graph(id="graph-3d-plot-tsne", style={"height": "98vh"})
+                        ],
+                    ),
+                    html.Div(
+                        className="three columns",
+                        id="euclidean-distance",
+                        children=[
+                            Card(
+                                style={"padding": "5px"},
+                                children=[
+                                    html.Div(
+                                        id="div-plot-click-message",
+                                        style={
+                                            "text-align": "center",
+                                            "margin-bottom": "7px",
+                                            "font-weight": "bold",
+                                        },
+                                    ),
+                                    html.Div(id="div-plot-click-image"),
+                                    html.Div(id="div-plot-click-wordemb"),
+                                ],
+                            )
+                        ],
+                    ),
+                ],
+            ),
+        ],
+    )
+
+
+def demo_callbacks(app):
+    def generate_figure_image(groups, layout):
+        data = []
+
+        for idx, val in groups:
+            scatter = go.Scatter3d(
+                name=idx,
+                x=val["x"],
+                y=val["y"],
+                z=val["z"],
+                text=[idx for _ in range(val["x"].shape[0])],
+                textposition="top center",
+                mode="markers",
+                marker=dict(size=3, symbol="circle"),
+            )
+            data.append(scatter)
+
+        figure = go.Figure(data=data, layout=layout)
+
+        return figure
+
+    # Scatter Plot of the t-SNE datasets
+    def generate_figure_word_vec(
+        embedding_df, layout, wordemb_display_mode, selected_word, dataset
+    ):
+
+        try:
+            # Regular displays the full scatter plot with only circles
+            if wordemb_display_mode == "regular":
+                plot_mode = "markers"
+
+            # Nearest Neighbors displays only the 200 nearest neighbors of the selected_word, in text rather than circles
+            elif wordemb_display_mode == "neighbors":
+                if not selected_word:
+                    return go.Figure()
+
+                plot_mode = "text"
+
+                # Get the nearest neighbors indices using Euclidean distance
+                vector = data_dict[dataset].set_index("0")
+                selected_vec = vector.loc[selected_word]
+
+                def compare_pd(vector):
```

Comment:
Why is `compare_pd` not declared outside of this function. Also, `compare_pd` is also used in line 576 and line 579. So I recommend just declaring `compare_pd` outside of this function's scope.

Reply:
Ignore this comment. `vector.apply()` takes in a function as a parameter, so the declaration within the function.